File: lua/coroutines/3.lua
package.path = package.path .. ";../?.lua"
require("common")

local events = {}
function log(event)
  events[#events + 1] = event
end

-- Yield/resume from coroutine subfunction

function subfunction()
  log("subfunction started")
  log(debug.traceback())
  coroutine.yield(42)
  log("subfunction finished")
end

local c = coroutine.create(function(n)
  log("coroutine started")
  subfunction()
  log("coroutine finished")
end)

log("main started")
_, n = coroutine.resume(c)
log("main received: " .. to_s(n))
_, n = coroutine.resume(c)
log("main received: " .. to_s(n))
log("main finished")


expect_to_be_equal(events, {
  [1] = "main started",
  [2] = "coroutine started",
  [3] = "subfunction started",
  [4] = "stack traceback:\n" ..
        "	3.lua:13: in function 'subfunction'\n" ..
        "	3.lua:20: in function <3.lua:18>",
  [5] = "main received: 42",
  [6] = "subfunction finished",
  [7] = "coroutine finished",
  [8] = "main received: nil",
  [9] = "main finished",
})

File: js/async-await/3.js
require("./common")();

let events = [];

function log(event) {
  events.push(event);
}

// Yield/resume from generator subfunction

function* subfunction() {
  let f = () => {
    // yield 42; // fails to run even though the outer function is generator
  }
  log("subfunction started");
  // log(stackTrace());
  yield 42; // can't yield without marking function as generator
  log("subfunction finished");
}

function* cc() {
  log("coroutine started");
  let sf = subfunction();
  yield sf.next().value;
  yield sf.next().value;
  log("coroutine finished");
}

let c = cc();
log("received: " + JSON.stringify(c.next()));
log("received: " + JSON.stringify(c.next()));
log("received: " + JSON.stringify(c.next()));

expectToEqual(events, [
  "coroutine started",
  "subfunction started",
  "received: {"value":42,"done":false}",
  "subfunction finished",
  "received: {"done":false}",
  "coroutine finished",
  "received: {"done":true}"
]);
